daml/Marketplace/Clearing/Service.daml

```diff
@@ -0,0 +1,212 @@
+          create MarkToMarketCalculation with ccp = provider; accountId = clearingAccount.id; ..
+
+      nonconsuming PerformMarginFill : Optional ((ContractId AssetDeposit), [ContractId AssetDeposit])
+        with
+          depositCids : [ContractId AssetDeposit]
+          marginDepositCids : [ContractId AssetDeposit]
+          calculationCid : ContractId MarginCalculation
+        do
+          targetAmount  <- (.targetAmount) <$> fetch calculationCid
+          depositAmount <- getDepositQuantities depositCids
+          marginAmount  <- getDepositQuantities marginDepositCids
+
+          let diff = targetAmount - marginAmount
+          if diff > 0.0 && depositAmount < diff
+          then resolveMarginCalculation calculationCid CalculationFailed >> return None
+          else do
+            resolveMarginCalculation calculationCid CalculationSatisfied
+            if | diff == 0.0 -> return None
+               | diff > 0.0  -> Some <$> exercise self TransferToMargin with amount = diff; ..
+               | otherwise   -> Some <$> exercise self TransferFromMargin with amount = (abs diff); ..
+
+      nonconsuming PerformMarkToMarket : Optional (ContractId AssetDeposit, [ContractId AssetDeposit])
+        with
+          providerDepositCids : [ContractId AssetDeposit]
+          customerDepositCids : [ContractId AssetDeposit]
+          calculationCid : ContractId MarkToMarketCalculation
+        do
+          mtmAmount      <- (.mtmAmount) <$> fetch calculationCid
+          providerAmount <- getDepositQuantities providerDepositCids
+          customerAmount <- getDepositQuantities customerDepositCids
+
+          let failCalc : Update ()
+              failCalc = resolveMarkToMarketCalculation calculationCid CalculationFailed
+
+              satisfyCalc : Update ()
+              satisfyCalc = resolveMarkToMarketCalculation calculationCid CalculationSatisfied
+
+          if
+            | mtmAmount == 0.0 -> satisfyCalc >> return None
+            | mtmAmount > 0.0  ->
+              if providerAmount < mtmAmount
+              then failCalc >> return None
+              else satisfyCalc >> Some <$> exercise self TransferFromProvider with amount = mtmAmount; depositCids = providerDepositCids
+            | otherwise ->
+              if customerAmount < abs mtmAmount
+              then failCalc >> return None
+              else satisfyCalc >> Some <$> exercise self TransferToProvider with amount = abs mtmAmount; depositCids = customerDepositCids
+
+      nonconsuming TransferFromProvider : (ContractId AssetDeposit, [ContractId AssetDeposit])
+        with
+          depositCids : [ContractId AssetDeposit]
+          amount : Decimal
+        do
+          assertMsg "deposits are not enough to transfer from provider" =<< (amount <=) <$> getDepositQuantities depositCids
+          (depositCid::remainingCids) <- mergeAndSplit depositCids amount
+          flip (,) remainingCids <$> transferWithController depositCid clearingAccount
```

Comment:
I had been looking for a way to have the `remainingCids` on the right hand side...that didn't work since `transferController` is in `Update` but I was able to do it with the following using `<&>` from `DA.Functor`:
`transferWithController depositCid clearingAccount <&> (, remainingCids)`, so thanks for the tip with the tuple construction syntax!

I will probably go with a new result type though.